core/services/workflows/engine.go

```diff
@@ -19,100 +22,143 @@ const (
 	mockedTriggerID   = "cccccccccc0000000000000000000000"
 )
 
+// Engine handles the lifecycle of a single workflow and its executions.
 type Engine struct {
 	services.StateMachine
-	logger     logger.Logger
-	registry   types.CapabilitiesRegistry
-	trigger    capabilities.TriggerCapability
-	consensus  capabilities.ConsensusCapability
-	targets    []capabilities.TargetCapability
-	workflow   *Workflow
-	callbackCh chan capabilities.CapabilityResponse
-	cancel     func()
+	logger              logger.Logger
+	registry            types.CapabilitiesRegistry
+	workflow            *workflow
+	executionStates     *inMemoryStore
+	pendingStepRequests chan stepRequest
+	triggerEvents       chan capabilities.CapabilityResponse
+	newWorkerCh         chan struct{}
+	stepUpdateCh        chan stepState
+	wg                  sync.WaitGroup
+	stopCh              services.StopChan
+	newWorkerTimeout    time.Duration
+
+	// Used for testing to wait for an execution to complete
+	xxxExecutionFinished chan string
 }
 
 func (e *Engine) Start(ctx context.Context) error {
 	return e.StartOnce("Engine", func() error {
 		// create a new context, since the one passed in via Start is short-lived.
-		ctx, cancel := context.WithCancel(context.Background())
-		e.cancel = cancel
+		ctx, _ := e.stopCh.NewCtx()
+
+		e.wg.Add(2)
 		go e.init(ctx)
-		go e.triggerHandlerLoop(ctx)
+		go e.loop(ctx)
+
 		return nil
 	})
 }
 
+// init does the following:
+//
+//  1. Resolves the underlying capability for each trigger
+//  2. Registers each step's capability to this workflow
+//  3. Registers for trigger events now that all capabilities are resolved
+//
+// Steps 1 and 2 are retried every 5 seconds until successful.
```

Comment:
Hmmm... probably not -- I think this should be a temporary thing, and if we're not able to initialize the workflow we should keep trying since the NOP has chosen to run the job.